cranelift/chaos/src/enabled.rs

```diff
@@ -0,0 +1,171 @@
+use std::{
+    fmt::Debug,
+    sync::{Arc, Mutex},
+};
+
+use arbitrary::{Arbitrary, Unstructured};
+
+struct ChaosEngineData {
+    /// # Safety
+    ///
+    /// This field must never be moved from, as it is referenced by
+    /// the field `unstructured` for its entire lifetime.
+    ///
+    /// This pattern is the ["self-referential" type](
+    /// https://morestina.net/blog/1868/self-referential-types-for-fun-and-profit)
```

Comment:
from a safety perspective, the other extremely important detail of this trick is that `ChaosEngineData` must also never move once references to `engine.data` are taken. so the "this must not move"-ness of `data` kind of percolates through to any enclosing type until it's somewhere that won't move (which works out here because `ChaosEngineData` ends up owned by an `Arc` where it oughtn't be moved out of.

as an example that certainly won't come up here but would be Technically Possible, `Arc::new(some_chaos_engine.data.try_unwrap())` would yield a `ChaosEngineData` whose `unstructured` points to somewhere else, and would (hopefully! :D) fault on use.

Reply:
Thanks for pointing that out! Does it also apply if some type is heap allocated? I think the article I got this from used a `Box` to create a level of indirection. The idea being that if the `Box` itself is moved, the values on the heap won't. So any existing references into that heap allocation would still be valid. In this case, the `Vec` is supposed to serves the same purpose as the `Box` in the article.

That being said, I just noticed that I got the order of the fields wrong, which the article warns against. `data` will be dropped before `unstructured`, which creates a dangling pointer and UB.(?) *oops* :smile: I definitely prefer a safe solution as well.

Reply:
Aaand reading a bit further, I also forgot the thing about `AliasableBox` so you're definitely right, moving `data` would also be UB.

Reply:
yeah that's the part that makes the blog post's solution a little more robust to `move`s - a `&AliasableBox<ZipArchive<File>>` could be made to dangle,  but with private internals you can ensure that wouldn't happen. anyway, hopefully threading a `&mut ControlPlane` through the compiler as appropriate lets you avoid the whole construction, and double-hopefully the extra arguments don't affect compile time all that much :)